Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py

```diff
@@ -444,6 +633,206 @@ def preview(self):
+      return
+
+    sliceLogic = self.histogramPipeline.sliceWidget.sliceLogic()
+    backgroundLogic = sliceLogic.GetBackgroundLayer()
+    self.reslice.SetInputConnection(backgroundLogic.GetReslice().GetInputConnection(0, 0))
+    self.reslice.SetResliceTransform(backgroundLogic.GetReslice().GetResliceTransform())
+    self.reslice.SetInterpolationMode(backgroundLogic.GetReslice().GetInterpolationMode())
+    self.reslice.SetOutputExtent(brushExtent)
+
+    maxNumberOfBins = 1000
+    masterImageData = self.scriptedEffect.masterVolumeImageData()
+    scalarRange = masterImageData.GetScalarRange()
+    numberOfBins = int(scalarRange[1] - scalarRange[0]) + 1
+    if numberOfBins > maxNumberOfBins:
+      numberOfBins = maxNumberOfBins
+    binSpacing = (scalarRange[1] - scalarRange[0] + 1) / numberOfBins
+
+    self.imageAccumulate.SetComponentExtent(0, numberOfBins - 1, 0, 0, 0, 0)
+    self.imageAccumulate.SetComponentSpacing(binSpacing, binSpacing, binSpacing)
+    self.imageAccumulate.SetComponentOrigin(scalarRange[0], scalarRange[0], scalarRange[0])
+
+    self.imageAccumulate.Update()
+
+    self.histogramFunction.RemoveAllPoints()
+    tableSize = self.imageAccumulate.GetOutput().GetPointData().GetScalars().GetNumberOfTuples()
+    for i in range(tableSize):
+      value = self.imageAccumulate.GetOutput().GetPointData().GetScalars().GetTuple1(i)
+      self.histogramFunction.AddPoint(binSpacing * i + scalarRange[0], value)
+    self.histogramFunction.AdjustRange(scalarRange)
+
+    lower  = self.imageAccumulate.GetMin()[0]
+    average = self.imageAccumulate.GetMean()[0]
+    upper = self.imageAccumulate.GetMax()[0]
+
+    # If there is a selection, then set the threshold based on that
+    if self.selectionStartPosition is not None and self.selectionEndPosition is not None:
+
+      # Clamp selection based on scalar range
+      startX = min(scalarRange[1], max(scalarRange[0], self.selectionStartPosition.x()))
+      endX = min(scalarRange[1], max(scalarRange[0], self.selectionEndPosition.x()))
+
+      lower = min(startX, endX)
+      average = (startX + endX) / 2.0
+      upper = max(startX, endX)
+
+    epsilon = 0.00001
+    self.minMaxFunction.RemoveAllPoints()
+    self.minMaxFunction.AddPoint(lower - epsilon, 0.0)
+    self.minMaxFunction.AddPoint(lower, 1.0)
+    self.minMaxFunction.AddPoint(lower + epsilon, 0.0)
+    self.minMaxFunction.AddPoint(upper - epsilon, 0.0)
+    self.minMaxFunction.AddPoint(upper, 1.0)
+    self.minMaxFunction.AddPoint(upper + epsilon, 0.0)
+    self.minMaxFunction.AdjustRange(scalarRange)
+
+    self.averageFunction.RemoveAllPoints()
+    self.averageFunction.AddPoint(average - epsilon, 0.0)
+    self.averageFunction.AddPoint(average, 1.0)
+    self.averageFunction.AddPoint(average + epsilon, 0.0)
+    self.averageFunction.AdjustRange(scalarRange)
+
+    minimumThreshold = lower
+    maximumThreshold = upper
+
+    if self.histogramLowerThresholdMinimumButton.checked:
```

Comment:
it would be a bit nicer to use parameter value in self.scriptedEffect node instead of GUI button state